policyengine_api/utils/logger.py

```diff
@@ -0,0 +1,116 @@
+import logging
+import sys
+from google.cloud import logging as cloud_logging
+from datetime import datetime
+from pathlib import Path
+import os
+
+
+class Logger:
+    def __init__(
+        self,
+        dir: str = "logs",
+        name: str = "api_main",
+        id: str = None,
+        log_to_cloud: bool = True,
+    ):
+        """
+        Initialize standard logger
+
+        Filepath:
+        dir/name_id.log
+
+        Args:
+            dir (str): Directory to store log files (defaults to "logs")
+            name (str): Name of the logger (defaults to "api_main")
+            id (str): ID to append to log file name; if not provided, will use current timestamp
+            log_to_cloud (bool): Whether to log to Google Cloud Logging
+        """
+        # Generate three parts of storage path
+        self.dir = Path(dir)
+        self.name = name
+        # Don't pass datetime to initializer, otherwise time is equal to when class is created
+        self.id = (
+            id if id is not None else datetime.now().strftime("%Y%m%d_%H%M%S")
+        )
+
+        self.logger = logging.getLogger(self.name)
+        self.logger.setLevel(logging.INFO)
+
+        # Create log directory if it doesn't exist
+        try:
+            self.dir.mkdir(parents=True, exist_ok=True)
+        except Exception as e:
+            print(
+                f"Warning: Could not create log directory {self.dir}: {str(e)}"
+            )
+            # Fall back to current directory
+            self.dir = Path(".")
+
+        # Create log file path based upon directory
+        self.filepath = self.dir.joinpath(f"{self.name}_{self.id}.log")
+
+        self.cloud_client: cloud_logging.Client = None
+
+        # Prevent duplicate handlers
+        if not self.logger.handlers:
+            # File handler - logs to local file
+            file_handler = logging.FileHandler(str(self.filepath))
+            file_handler.setLevel(logging.INFO)
+            file_handler.setFormatter(
+                logging.Formatter(
+                    "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
+                )
+            )
+            self.logger.addHandler(file_handler)
+
+            # Console handler
+            console_handler = logging.StreamHandler(sys.stdout)
+            console_handler.setLevel(logging.INFO)
+            console_handler.setFormatter(
+                logging.Formatter(
+                    "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
+                )
+            )
+            self.logger.addHandler(console_handler)
+
+            # Google Cloud Logging handler; don't log to GCP if in debug
+            if log_to_cloud and os.environ.get("FLASK_DEBUG") != "1":
```

Comment:
Similarly you could configure this at the application level rather than in code at each logger declaration. Maybe there are reasons to only cloud log some things, but I'd generally expect it to be on/off for the whole app.